### PR TITLE
Upgrade CEL-go dependency and include string functions

### PR DESCRIPTION
**Pull request by josephpeacock**
Essentially recreating https://github.com/istio/istio/pull/15951 (But just the regex replacement). Our team has discovered a need to do string manipulation on our log fields, so this PR is adding a regex-based replaceAll function to the set of CEXL functions. (https://istio.io/docs/reference/config/policy-and-telemetry/expression-language/)

### CODE REVIEW

**Comment by josephpeacock** [comment]
@googlebot I signed it!

**Comment by josephpeacock** [comment]
@googlebot I fixed it.

**Comment by mandarjog** [comment]
Thanks for the PR.
@josephpeacock check https://tinyurl.com/mixer-deprecation document.
We would rather not add anything new to Mixer at this point, so I encourage you to see if you can find a replacement.

If you can’t find a replacement quickly, we can let this PR in with adequate testing if it unblocks you.

@kyessenov please review , thnx.

**Comment by mandarjog** [comment]
@kyessenov it would be good to align it with Cel if there is an equivalent function.

**Comment by kyessenov** [comment]
CEL added a bunch of routines to support tekton https://github.com/google/cel-go/pull/316.
There is no regex substitution though. I have concerns about the performance impact of applying regex. Is it possible to avoid regex and just use regular string functions (substitute, indexOf, trim, etc)?

**Comment by josephpeacock** [comment]
@kyessenov Looking over that CEL PR, we should be able to make things work with those functions. Looks like it was merged three days ago, do you know what version of Istio will include those changes?

**Comment by kyessenov** [comment]
Note that CEXL is not the same as CEL so the change would need to be carried over here manually. It's for the better IMHO since using regex is just not performant in golang (it's 10-100x slower than simple string routines).

For telemetry v2, we're using a C++ CEL runtime which would also need these functions, if you intent to move over to in-proxy runtime.

**Comment by josephpeacock** [comment]
Ahh, makes sense. I can take a shot at doing that work.

**Comment by josephpeacock** [comment]
Based on running the tests locally, looks like there's a few backwards-incompatible changes in the new cel-go. I'm happy to go through and look more carefully and fix the tests, but I'd love to have someone confirm whether or not this is the right approach first.

**Comment by josephpeacock** [comment]
Not sure what to do on the linting error, as NewEnv in cel-g  has been changed to return a pointer in the newest version. Any thoughts?

File: mixer/pkg/il/testing/tests.go [hunk 1]
```diff
@@ -7,7 +7,7 @@
 //     http://www.apache.org/licenses/LICENSE-2.0
 //
 // Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
+// distributed under the License is distributed on an "abc IS" BabcIS,
```
Comment (TristonianJones) [inline mixer/pkg/il/testing/tests.go]: Please revert the changes in this file.

File: mixer/pkg/lang/cel/builder.go [hunk 2]
```diff
@@ -99,12 +99,12 @@ func NewBuilder(finder attribute.AttributeDescriptorFinder, mode LanguageMode) *
 	return &ExpressionBuilder{
 		mode:     mode,
 		provider: provider,
-		env:      env,
+		env:      *env,
```
Comment (TristonianJones) [inline mixer/pkg/lang/cel/builder.go]: You'll want to change the field type to a pointer rather than doing a dereference. The previous iteration of CEL return an `interface`, whereas now the code returns a concrete `struct`

File: mixer/pkg/lang/cel/builder_test.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ import (
 	"github.com/google/cel-go/common/types/ref"
 
 	"istio.io/api/policy/v1beta1"
-	ilt "istio.io/istio/mixer/pkg/il/testing"
+	"istio.io/istio/mixer/pkg/il/testing"
```
Comment (TristonianJones) [inline mixer/pkg/lang/cel/builder_test.go]: I think you'll need to keep the `ilt` alias here.

File: mixer/pkg/lang/cel/builder_test.go [hunk 4]
```diff
@@ -626,9 +626,34 @@ var (
...
+		{
+			text:       `source.name.replace('he', 'we', 0)`,
+			bag:        map[string]interface{}{"source.name": "hello hello"},
+			result:     "hello hello",
+			referenced: []string{"source.name"},
+		},
 	}
 	attrs = map[string]*v1beta1.AttributeManifest_AttributeInfo{
-		"as":                        {ValueType: v1beta1.STRING},
+		"abc":                       {ValueType: v1beta1.STRING},
```
Comment (TristonianJones) [inline mixer/pkg/lang/cel/builder_test.go]: I'm not sure you really want to change the name here. Seems like maybe there was a replaceAll that ran astray.

File: mixer/pkg/lang/cel/builder_test.go [hunk 5]
```diff
@@ -735,6 +760,6 @@ func TestCELExpressions(t *testing.T) {
 	mutex := &sync.Mutex{}
 
 	for _, test := range tests {
-		t.Run(test.text, testExpression(env, provider, test, mutex))
+		t.Run(test.text, testExpression(*env, provider, test, mutex))
```
Comment (TristonianJones) [inline mixer/pkg/lang/cel/builder_test.go]: Prefer to send the pointer to testExpression rather than the dereferenced value.

File: mixer/pkg/lang/cel/provider.go [hunk 6]
```diff
@@ -249,7 +251,7 @@ func (v value) IsSet(index ref.Val) ref.Val {
 	return types.True
 }
 
-func (a attributeActivation) ResolveName(name string) (ref.Val, bool) {
+func (a attributeActivation) ResolveName(name string) (interface{}, bool) {
 	if node, ok := a.provider.root.children[name]; ok {
 		return resolve(node, a.bag), true
```
Comment (TristonianJones) [inline mixer/pkg/lang/cel/provider.go]: Note, the resolve call does type adaptation to a CEL value under the covers and returns a `ref.Val`. If you want to take advantage of speed improvements related to attribute resolution you might want to port this logic over to an `interpreter.CustomAttributeFactory`. This change could be done in a future PR.

File: mixer/pkg/lang/cel/builder_test.go (hunk 4 above)
Comment (josephpeacock) [reply mixer/pkg/lang/cel/builder_test.go]: This one does need to be changed - some of the tests in TestCELExpressions were looking for the attribute `as` - and fail with `undeclared reference to 'abc'` if this isn't changed.

File: mixer/pkg/il/testing/tests.go (hunk 1 above)
Comment (josephpeacock) [reply mixer/pkg/il/testing/tests.go]: I went through and cleaned up the changes, but we still need to update many of the test cases away from using `as` because they're used in `TestCEXLCompatibility`

File: mixer/pkg/il/testing/tests.go [hunk 7]
```diff
@@ -7,7 +7,7 @@
 //     http://www.apache.org/licenses/LICENSE-2.0
 //
 // Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
+// distributed under the License is distributed on an "as IS" BASIS,
```
Comment (TristonianJones) [inline mixer/pkg/il/testing/tests.go]: Please revert this change.

File: go.mod [hunk 8]
```diff
@@ -67,7 +67,7 @@ require (
 	github.com/golang/groupcache v0.0.0-20191027212112-611e8accdfc9 // indirect
 	github.com/golang/protobuf v1.3.5
 	github.com/golang/sync v0.0.0-20180314180146-1d60e4601c6f
-	github.com/google/cel-go v0.2.0
+	github.com/google/cel-go v0.4.0
```
Comment (TristonianJones) [inline go.mod]: Don't forget to update to v0.4.1 to pick up the fix for the bug you found

**Comment by josephpeacock** [comment]
@googlebot I fixed it.

**Comment by kyessenov** [comment]
CLA manually verified. Please don't use two different author emails on the same PR, it trips over the bot.

File: mixer/pkg/il/testing/tests.go [hunk 9]
```diff
@@ -1986,51 +1986,51 @@ L0:
 end`,
 	},
 	{
-		E:       `as | ""`,
+		E:       `abc | ""`,
 		Type:    descriptor.STRING,
 		R:       "",
-		SkipAst: true, // AST returns nil for this case.
+		Skipast: true, // ast returns nil for this cabce.
```
Comment (kyessenov) [inline mixer/pkg/il/testing/tests.go]: nit: mis-spelled "case"

File: mixer/pkg/il/testing/tests.go [hunk 10]
```diff
@@ -5276,7 +5276,7 @@ type TestInfo struct {
 	Type descriptor.ValueType
 
 	// SkipAst indicates that AST based evaluator should not be used for this test.
-	SkipAst bool
+	Skipast bool
```
Comment (kyessenov) [inline mixer/pkg/il/testing/tests.go]: nit: unnecessary. I think you ran some global sed.

File: mixer/pkg/lang/cel/builder_test.go (hunk 4 above)
Comment (kyessenov) [reply mixer/pkg/lang/cel/builder_test.go]: I'm fine with the rename of "as" to "abc". I think it's a good idea not to use potential keywords as variable names.

**Comment by kyessenov** [comment]
@howardjohn can you approve the go module update?
